### DISCUSSION TITLE
Add `has_add_on` to the check-in boolean logic

### DISCUSSION THREAD

**Discussion by kiancross** [pretix/pretix#3523] [Ideas]
We have various ticket configurations, with some providing access to different ticket gates. This should soon work great if each ticket is a separate product (https://github.com/pretix/pretix/discussions/3466), but not so much if these tickets are configured using add-on products (e.g., purchasing an upgrade for your ticket, which allows you to enter through a fast-track gate).

I propose a `has_add_on` selector in the check-in boolean logic, which checks whether a product has a particular add-on associated with it.

Any thoughts?

# Motivation

It may be worth stating the motivation for this. Suppose we have numerous base tickets of varying costs. Ticket holders can purchase a set of upgrades (add-on products, which cost the same regardless of the base ticket). These add-on products provide access to different entrances (as well as other perks).

If we were to do this using only products, we would need to create a product variation for each combination of base tickets and add-on products. This is ok for a couple of base tickets and a couple of add-ons, but the number of variations grows quadratically, so it becomes tiresome to manage.

**Comment by raphaelm** [comment]
We have previously discussed this approach as one of two options for this situation, and decided against this approach. I think the final conversation happened internally, but much of the reasoning is in #2539.

What you can do is create a separate check-in list just for the add-ons and enable the addon_match flag.

It may not solve *all* theoretical use cases, but most, and it has the advantage of the check-in being stored with the add-on, not the main product. The use-case of a "upgrade for fast-track" is probably not solvable and would require a product variation, but for something like "here's a different part of the event that costs extra" it works just fine.

**Reply by kiancross** [reply]
I think the problem from https://github.com/pretix/pretix/discussions/3466 remains with this solution.

1. The ticket bundle permits one entry to the event, with add-ons simply changing the entry point. With this solution, because the add-on is checked in, the customer could send their ticket to somebody else to also use at the standard entry. I suppose I could write a custom plugin to create a check-in for the base ticket whenever an add-on is checked in, which would resolve this.
2. Some entrances can be used based on more than one add-on. And customers may have more than one of those add-ons. But the current logic does not allow ambiguous check-ins.

I think `has_add_on` is the only 'nice' solution to this problem. I think that combined with the `gate` selector which you are adding in https://github.com/pretix/pretix/pull/3521, this would allow very powerful check-in logic!

**Reply by raphaelm** [reply]
Yes, the "there are multiple entrances to the same part of the event which are controlled by add-ons" can not be solved by this. (But it's also a situation I'm hearing about for the first time right now, after doing all this for 9 years, so I don't think it's a very common problem to have.)

I'm not saying a definitive "no" to `has_add_on`, but I'm also not sold on it being worth the complexity for now.

**Reply by kiancross** [reply]
If you agreed to a hook in the check-in logic, then this could probably be solved with https://github.com/pretix/pretix/discussions/3524.
